decentraland/unity-explorer
# fix: sdk video player SEEK + VS_ERROR video event

### WHY

Today if the video player is initialized with a specific `position` value, the real position seeked by the video player library is not exactly the one expected.

Also, the `VS_ERROR` video event wasn't being used at all, when trying to load an invalid video source (unexistent local video / unreachable streaming url) that video event should be propagated to the scene.

### WHAT

* After reading the documentation comparing [`SeekFast()`](https://www.renderheads.com/content/docs/AVProVideo/api/RenderHeads.Media.AVProVideo.IMediaControl.html#RenderHeads_Media_AVProVideo_IMediaControl_SeekFast_System_Double_) and [`Seek()`](https://www.renderheads.com/content/docs/AVProVideo/api/RenderHeads.Media.AVProVideo.IMediaControl.html#RenderHeads_Media_AVProVideo_IMediaControl_Seek_System_Double_) methods, seeing that `Seek()` might take a bit more to buffer but takes the video to the exact expected time and `SeekFast()` takes the video to the closest frame from that other time, with less buffering; When testing both I see a better result with `Seek()`.
* Added a minimum wait until `seekableTimes` has at least 1 value during video initialization, to fix the first video play with seek.
* For MacOS the only way I found to make seek initialization work correctly **after reloading the scene** (e.g. hot-reload) is to have a 1 second wait before applying the seeking.
* Added `403: Forbidden Access` web request error contemplation as we need the unreachable streaming videos to have `VS_ERROR` video event and state ([here](https://github.com/decentraland/unity-explorer/pull/2563/commits/c317b1f2a8a501031db64b16303c63f126f3e036))
* Improved `MediaPlayerComponent.State` setting to make sure the `LastStateChangeTime` gets updated correctly

**NOTE:** I tried using [AVPRO's Events system](https://www.renderheads.com/content/docs/AVProVideo/articles/feature-events.html) for handling the errors, but I discovered it never throws any event at all. Also, the `mediaPlayerControl.GetLastError()` never has any error at all, even when AVPro fails to load the video...

Solves https://github.com/decentraland/creator-hub/issues/247, https://github.com/decentraland/creator-hub/issues/242 and other untracked video issues reported by the content team.

### QA INSTRUCTIONS (WinOS + MacOS)

#### Test setup

1. Download [this scene](https://github.com/user-attachments/files/17555919/video-event-seek-test_SEEK-ERRORS.zip) and uncompress it somewhere
2. Enter the scene root folder and run `npm i` and then `npm run start -- --explorer-alpha`
3. Close the Explorer that auto-opened. Leave the scene running in the console/terminal.
4. Download the build from this PR and connect it to the running scene using [the console command](https://github.com/decentraland/unity-explorer/wiki/How-to-connect-to-a-local-scene-(Unity-Editor---Custom-Build---Latest-Released-Build)#connecting-a-custom-build-to-the-scene) according to your OS
5. Confirm the expected behaviours described below
6. **Additionally please do a smoke test with this PR build on other deployed scenes with video playing**

The test consists in running a provided test scene and confirming the expected behaviours.

The test scene has a video screen and a **clickable cube** that every time it's clicked, it changes to the next video in a collection of 3 video sources:
0 -> `https://player.vimeo.com/external/677209284.m3u8?s=9dd022d3f601a898b57eef7ea7b43ef1f2e0cf57eef` **[INVALID VIDEO]**
1 -> `https://player.vimeo.com/external/677209284.m3u8?s=9dd022d3f601a898b57eef7ea7b43ef1f2e0cf56` **[VALID VIDEO]**
2 -> `fedwvewf` **[INVALID VIDEO]**

#### `VS_ERROR` Expected behaviour
When the UI shows that the video is `0` or `2` the video event displayed in the UI should end up being `VS_ERROR` (it can take a little time before)

When the UI shows that the video is `1` the video event should not display `VS_ERROR` and should end up loading the video and showing the `VS_PLAYING` video event in the UI.

#### SEEK consistent frame

Swap between the videos getting to the video `1` several times, waiting for it to play and when it plays confirm that it always starts around this frame:

<img width="1175" alt="Screenshot 2024-10-29 at 11 57 11 AM" src="https://github.com/user-attachments/assets/c9d71597-abd4-4498-9109-48c39efde0df">

#### Video demonstration

https://github.com/user-attachments/assets/98dd47d4-9df2-488f-91ad-56b117f61d2e
